### ISSUE TITLE
SSL: CERTIFICATE_VERIFY_FAILED when build with pyinstaller + latest dependencies 

### ISSUE THREAD

**Issue by robbinvandamme** [Azure/azure-iot-sdk-python#991]
# Context

- **OS and version used:** EmbeddedLinx kernel 5.10.65 armv7l GNU/Linux
- **Python version:** python3.7
- **pip version:** 19.0.3
- **list of installed packages:**:
altgraph==0.17.2
azure-iot-device==2.11.0
certifi==2021.10.8
charset-normalizer==2.0.12
deprecation==2.1.0
idna==3.3
importlib-metadata==4.11.3
janus==1.0.0
packaging==21.3
paho-mqtt==1.6.1
pyinstaller==5.0
**pyinstaller-hooks-contrib==2022.4**
pyparsing==3.0.8
PySocks==1.7.1
requests==2.27.1
requests-unixsocket==0.3.0
typing-extensions==4.2.0
urllib3==1.26.9
zipp==3.8.0

## Description of the issue

We are using this SDK in our application, which is build with pyinstaller running om embedded linux.
After upgrading all our requirements we ran into SSL: CERTIFACATE_VERIFY_FAILED error when build it with pyinstaller (and his latest dependencies) and running it on our target.
After some digging I found that a newer version of a dependency of pyinstaller caused the issue, being pyinstaller-hooks-contrib.
pyinstaller-hooks-contrib==2021.3 is still ok, but starting from pyinstaller-hooks-contrib==2021.4 things go south.

If you check the changelog of this pip module:
https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/master/CHANGELOG.rst#updated-hooks-5
There they mention:
Removed the certifi run-time hook because it was not required for certifi to function in a frozen application. It was sometimes setting the SSL_CERT_FILE environment variable which causes applications to behave differently when frozen. In particular the SSLContext.set_default_verify_paths() method loads the certificates from certifi when the SSL_CERT_FILE environment variable is set. (https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues/335)

Then found this related ticket:
https://github.com/pyinstaller/pyinstaller/issues/6667
The maintainer suggest setting SSL_CERT_FILE. --> https://github.com/pyinstaller/pyinstaller/issues/6667#issuecomment-1064316393
But this certificate comes from/with the SDK? So the SDK should point to it via this env variable?

However, we can continue by setting this pip module fixed to 2021.3, but this is just a temporary fix.
Going forward I think a solution must be found in order to keep using this SDK together with pyinstaller (and his latest dependencies) on our (and other) embedded IoT-devices.

## Code sample exhibiting the issue

await iot_hub_device_client.connect()

## Console log of the issue
2022-04-27T10:41:37.464Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(104): wrapper: Starting run_op in pipeline thread
2022-04-27T10:41:37.466Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(82): _get_named_executor: Creating pipeline executor
2022-04-27T10:41:37.471Z azure.iot.device.common.pipeline.pipeline_stages_http DEBUG pipeline_stages_http(50): _run_op: Gateway Hostname not present. Setting Hostname to: renson-iot-dev-iotcloudcore-ioth.azure-devices.net
2022-04-27T10:41:37.475Z azure.iot.device.common.pipeline.pipeline_stages_http DEBUG pipeline_stages_http(56): _run_op: HTTPTransportStage(InitializePipelineOperation): got connection args
2022-04-27T10:41:37.478Z azure.iot.device.common.http_transport DEBUG http_transport(67): _create_ssl_context: creating a SSL context
2022-04-27T10:41:37.486Z azure.iot.device.common.pipeline.pipeline_ops_base DEBUG pipeline_ops_base(111): complete: InitializePipelineOperation: completing without error
2022-04-27T10:41:37.488Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(104): wrapper: Starting <azure.iot.device.common.evented_callback.EventedCallback object at 0xb5bfee90> in callback thread
2022-04-27T10:41:37.492Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(82): _get_named_executor: Creating callback executor
2022-04-27T10:41:37.496Z azure.iot.device.common.evented_callback DEBUG evented_callback(50): wrapping_callback: Callback completed with result None
2022-04-27T10:41:37.502Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(104): wrapper: Starting run_op in pipeline thread
2022-04-27T10:41:37.506Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(460): _start_token_update_alarm: SasTokenStage: Scheduling automatic SAS Token renewal at epoch time: 1651052377
2022-04-27T10:41:37.514Z azure.iot.device.common.pipeline.pipeline_stages_mqtt DEBUG pipeline_stages_mqtt(126): _run_op: Gateway Hostname not present. Setting Hostname to: renson-iot-dev-iotcloudcore-ioth.azure-devices.net
2022-04-27T10:41:37.516Z azure.iot.device.common.pipeline.pipeline_stages_mqtt DEBUG pipeline_stages_mqtt(132): _run_op: MQTTTransportStage(InitializePipelineOperation): got connection args
2022-04-27T10:41:37.519Z azure.iot.device.common.mqtt_transport DEBUG mqtt_transport(136): _create_mqtt_client: creating mqtt client
2022-04-27T10:41:37.521Z azure.iot.device.common.mqtt_transport INFO mqtt_transport(149): _create_mqtt_client: Creating client for connecting using MQTT over TCP
2022-04-27T10:41:37.524Z azure.iot.device.common.mqtt_transport DEBUG mqtt_transport(320): _create_ssl_context: creating a SSL context
2022-04-27T10:41:37.528Z azure.iot.device.common.mqtt_transport DEBUG mqtt_transport(327): _create_ssl_context: configuring SSL context with default certs
2022-04-27T10:41:37.531Z azure.iot.device.common.mqtt_transport DEBUG mqtt_transport(277): _create_mqtt_client: Created MQTT protocol client, assigned callbacks
2022-04-27T10:41:37.533Z azure.iot.device.common.pipeline.pipeline_ops_base DEBUG pipeline_ops_base(111): complete: InitializePipelineOperation: completing without error
2022-04-27T10:41:37.536Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(104): wrapper: Starting <azure.iot.device.common.evented_callback.EventedCallback object at 0xb5bd4230> in callback thread
2022-04-27T10:41:37.539Z azure.iot.device.common.evented_callback DEBUG evented_callback(50): wrapping_callback: Callback completed with result None
2022-04-27T10:41:37.543Z azure.iot.device.iothub.aio.loop_management DEBUG loop_management(38): _make_new_loop: Creating new event loop - CLIENT_INTERNAL_LOOP
2022-04-27T10:41:37.546Z asyncio DEBUG selector_events(53): __init__: Using selector: EpollSelector
2022-04-27T10:41:37.568Z root INFO conn_test(59): main: after create_from_connection_string
2022-04-27T10:41:37.571Z root INFO conn_test(63): main: Run forever!
2022-04-27T10:41:37.574Z azure.iot.device.iothub.aio.async_clients INFO async_clients(226): connect: Connecting to Hub...
2022-04-27T10:41:37.578Z azure.iot.device.iothub.pipeline.mqtt_pipeline DEBUG mqtt_pipeline(254): connect: Starting ConnectOperation on the pipeline
2022-04-27T10:41:37.582Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(104): wrapper: Starting run_op in pipeline thread
2022-04-27T10:41:37.586Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(1131): _run_op: ReconnectStage(ConnectOperation): State changes DISCONNECTED -> CONNECTING. Sending op down
2022-04-27T10:41:37.590Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(695): _block: ConnectionLockStage(ConnectOperation): blocking
2022-04-27T10:41:37.593Z azure.iot.device.common.pipeline.pipeline_stages_mqtt DEBUG pipeline_stages_mqtt(177): _run_op: MQTTTransportStage(ConnectOperation): connecting
2022-04-27T10:41:37.601Z azure.iot.device.common.pipeline.pipeline_stages_mqtt DEBUG pipeline_stages_mqtt(67): _start_connection_watchdog: MQTTTransportStage(ConnectOperation): Starting watchdog
2022-04-27T10:41:37.606Z azure.iot.device.common.mqtt_transport DEBUG mqtt_transport(378): connect: connecting to mqtt broker
2022-04-27T10:41:37.609Z azure.iot.device.common.mqtt_transport INFO mqtt_transport(389): connect: Connect using port 8883 (TCP)
2022-04-27T10:41:37.659Z azure.iot.device.common.mqtt_transport INFO mqtt_transport(292): _force_transport_disconnect_and_cleanup: Forcing paho disconnect to prevent it from automatically reconnecting
2022-04-27T10:41:37.663Z azure.iot.device.common.mqtt_transport DEBUG mqtt_transport(314): _force_transport_disconnect_and_cleanup: Done forcing paho disconnect
2022-04-27T10:41:37.665Z azure.iot.device.common.pipeline.pipeline_stages_mqtt INFO pipeline_stages_mqtt(191): _run_op: transport.connect raised error
2022-04-27T10:41:37.673Z azure.iot.device.common.pipeline.pipeline_stages_mqtt INFO pipeline_stages_mqtt(192): _run_op: Traceback (most recent call last):
  File "azure/iot/device/common/mqtt_transport.py", line 391, in connect
  File "paho/mqtt/client.py", line 914, in connect
  File "paho/mqtt/client.py", line 1073, in reconnect
  File "ssl.py", line 1117, in do_handshake
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "azure/iot/device/common/pipeline/pipeline_stages_mqtt.py", line 189, in _run_op
  File "azure/iot/device/common/mqtt_transport.py", line 403, in connect
azure.iot.device.common.transport_exceptions.TlsExchangeAuthError

2022-04-27T10:41:37.676Z azure.iot.device.common.pipeline.pipeline_stages_mqtt DEBUG pipeline_stages_mqtt(104): _cancel_connection_watchdog: MQTTTransportStage(ConnectOperation): cancelling watchdog
2022-04-27T10:41:37.678Z azure.iot.device.common.pipeline.pipeline_ops_base DEBUG pipeline_ops_base(109): complete: ConnectOperation: completing with error 
2022-04-27T10:41:37.682Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(674): on_operation_complete: ConnectionLockStage(ConnectOperation): op failed.  Unblocking queue with error: 
2022-04-27T10:41:37.684Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(704): _unblock: ConnectionLockStage(ConnectOperation): unblocking and releasing queued ops.
2022-04-27T10:41:37.687Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(708): _unblock: ConnectionLockStage(ConnectOperation): processing 0 items in queue for error=
2022-04-27T10:41:37.690Z azure.iot.device.common.pipeline.pipeline_stages_base DEBUG pipeline_stages_base(1345): on_complete: ReconnectStage(ConnectOperation): failed, state change CONNECTING -> DISCONNECTED
2022-04-27T10:41:37.693Z azure.iot.device.common.pipeline.pipeline_thread DEBUG pipeline_thread(104): wrapper: Starting on_complete in callback thread
2022-04-27T10:41:37.697Z azure.iot.device.common.async_adapter INFO async_adapter(72): wrapping_callback: Callback completed with error 
2022-04-27T10:41:37.704Z azure.iot.device.common.async_adapter INFO async_adapter(73): wrapping_callback: ['azure.iot.device.common.transport_exceptions.TlsExchangeAuthError\n']
2022-04-27T10:41:37.708Z root INFO conn_test(38): init: ClientError: Error in the IoTHub client due to TLS exchanges.
Traceback (most recent call last):
  File "azure/iot/device/common/mqtt_transport.py", line 391, in connect
  File "paho/mqtt/client.py", line 914, in connect
  File "paho/mqtt/client.py", line 1073, in reconnect
  File "ssl.py", line 1117, in do_handshake
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "azure/iot/device/iothub/aio/async_clients.py", line 33, in handle_result
  File "azure/iot/device/common/async_adapter.py", line 91, in completion
  File "azure/iot/device/common/pipeline/pipeline_stages_mqtt.py", line 189, in _run_op
  File "azure/iot/device/common/mqtt_transport.py", line 403, in connect
azure.iot.device.common.transport_exceptions.TlsExchangeAuthError

The above exception was the direct cause of the following exception:

**Comment by Mikkolehtimaki** [comment]
I get a similar error when connecting to a (transparent) gateway running IoT Edge Runtime 1.2. Everything works if I connect to runtime 1.1 hub. 

Here's my stacktrace:

```
INFO:azure.iot.device.common.mqtt_transport:Creating client for connecting using MQTT over TCP
INFO:azure.iot.device.iothub.aio.async_clients:Connecting to Hub...
INFO:azure.iot.device.common.mqtt_transport:Connect using port 8883 (TCP)
INFO:azure.iot.device.common.mqtt_transport:Forcing paho disconnect to prevent it from automatically reconnecting
INFO:azure.iot.device.common.pipeline.pipeline_stages_mqtt:transport.connect raised error
INFO:azure.iot.device.common.pipeline.pipeline_stages_mqtt:Traceback (most recent call last):
  File "/.../.venv/lib/python3.10/site-packages/azure/iot/device/common/mqtt_transport.py", line 390, in connect
    rc = self._mqtt_client.connect(
  File "/.../.venv/lib/python3.10/site-packages/paho/mqtt/client.py", line 914, in connect
    return self.reconnect()
  File "/.../.venv/lib/python3.10/site-packages/paho/mqtt/client.py", line 1073, in reconnect
    sock.do_handshake()
  File "/.../.pyenv/versions/3.10.2/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 1341, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for '<hostname>'. (_ssl.c:997)
```

The certificates are fine and I can verify them using openssl.

**Comment by cartertinney** [comment]
@robbinvandamme

I've taken a look into the issue with PyInstaller - it appears the change they made is correct - what was happening before was that PyInstaller was (incorrectly) always using the `certifi` certificates, even when it wasn't supposed to. The correct behavior is to use the default system certificates rather than the ones in `certifi`.

My guess is that if your application no longer works now that PyInstaller has fixed this bug, that would mean that your system's certificate stores are not correctly configured. Can you check your system configuration?

We would expect that certificates would load upon calling `SSLContext.load_default_certs()` from the `ssl` module in Python. Please try the following code in a Python session to validate that your certificates are loadable:

```python
import ssl

ctx = ssl.SSLContext()
ctx.load_default_certs()
print(ctx.get_ca_certs())
```

If this prints out an empty list, then your certs have not been loaded correctly.

**Comment by robbinvandamme** [comment]
@cartertinney 
I tried your sample code to load the certificates on:
* our embedded target
* a stock beaglebone
* an ubuntu20 laptop
* an ubuntu22 laptop
All with the same result, an empty list. 

I also create an executable for ARM and x86.
The x86 one works on my laptop with the latest pip packages.
The ARM one works on a stock beaglebone, but not on our embedded target.
On our embedded target I can only get it to work with the older version of the pyinstaller-hooks-contrib.

I you have any ideas, I'm glad to hear them, think I need to dive deeper in our embedded device and check why its not working with that latest pip module.

**Comment by robbinvandamme** [comment]
@cartertinney 
I printed the default ca certs from inside the sdk mqtt_transport and even that returns an empty list.
But connection to IoTHub is working since its on my laptop.

![image](https://user-images.githubusercontent.com/13780546/166650484-d34fa89a-2789-4757-b010-a68605fdcb96.png)

So can you give me some more info about how the sdk uses or handles the certificates?

On my embedded device, I see that (the SDK?) downloads a certificate. Is it that one that is used by the SDK?

![image](https://user-images.githubusercontent.com/13780546/166650941-a5ef0deb-8256-4108-9961-95ec8d530add.png)

**Comment by robbinvandamme** [comment]
@cartertinney I found the problem.

I was missing a symlink /usr/lib/ssl/certs that points to /etc/ssl/certs where the certificates are installed.
After adding that, connection with IoTHub using latest pyinstaller was ok.

So when using the latest pyinstaller, the certifi certs are not used anymore by default and it looks at the system certificates.
If we want to use the certifi cacert.pem, it we should set the environment variable our-self like this.

``` 
    cert_file = None
    if getattr(sys, 'frozen', False) and hasattr(sys, "_MEIPASS"):
        print(sys._MEIPASS)
        cert_file = sys._MEIPASS + "/certifi/cacert.pem"
        os.environ["SSL_CERT_FILE"] = cert_file
        print("cert_file: %s" % cert_file)
```

I've haven't decided what the best option is for us.
Whats your opinion about this? Set SSL_CERT_FILE and use certifi cacert.pem or not set it and rely on system installed certificates?

**Comment by robbinvandamme** [comment]
@cartertinney 
After some small tests I also think requests always uses the certifi certs by default. Because if I remove the /tmp/_MEIGGpP4m/certifi/cacert.pem while the application (build with pyinstaller) is running, it fails with error that it doesn't find it anymore.

If I run it without creating a pyinstaller package, it uses: ~/venv/lib/python3.8/site-packages/certifi/cacert.pem

Your SDK or our libssl was not able to find our system certs in /etc/ssl/certs by default, we had to create a symlink from /usr/lib/ssl/certs to it. After that everything seems to be fine.

In the past your SDK (or our application) indeed probably was working by accident on our system because SSL_CERT_FILE was set to the certifi cacert.pem by pyinstaller-hooks-contrib, and now that's not the case anymore.

If your SDK by default wants to use the system certs, I guess then it's better to just leave it this way instead of me forcing the SDK to use the certifi cacert.pem? In the end, all I had to fix was one symlink.

What's the reason the SDK is using the system certs by default and not the certifi certs? If the request module is doing it, there must be some kind of reason for it maybe? And the cacert.pem will always be there because requests is a dependency of the SDK en certifi is a dependency for requests.

**Comment by cartertinney** [comment]
Yes, that's what I was saying. The `requests` library, by default, will use `certifi` certs. However, our `azure-iot-device` library does not. This behavior of using the system store certs is universal across languages of the Azure IoT Device SDK. `requests` is currently a dependency of our SDK, but is only used for a very small number of tasks, and ideally may eventually be removed. Additionally where it is used in our SDK, it is configured to use the system store rather than `certifi`

**Comment by robbinvandamme** [comment]
@cartertinney this part is not clear to me:
"Additionally where it is used in our SDK, it is configured to use the system store rather than certifi"

I created a venv and installed your sdk.
That comes with "./venv_cert_test/lib/python3.7/site-packages/certifi/cacert.pem".
If I remove that cacert.pem and try my sample app or in fact just do "from azure.iot.device.aio import IoTHubDeviceClient", it fails...
So is still requires ./certifi/cacert.pem to be there.

Maybe it's not used after (re-)configuring the SDK or the way requests should use certificates, but it seems that it needs to be there on starting the application or importing all the modules?

I attached the traceback in a file. [traceback.txt](https://github.com/Azure/azure-iot-sdk-python/files/8629264/traceback.txt)

Probably expected behavior? 

Apart from that no further questions anymore :-D.
Big thanks for your time and explanation!
